config/routes.rb

```diff
@@ -113,6 +113,8 @@
   end
 
   resources :users, only: %i[index edit update]
+  resource :user_upgrade_request, only: %i[new show create], path_names: { new: "/" }, path: "/upgrade"
```

Comment:
Yeah, I've tied myself in knots around this one.

When we started this project I was much more used to writing small javascript apps in express, where i really like explicit routing and handlers - it makes it clear exactly whats mounted on URLs. You can see that reflected in our routes file.

It's been pointed out to me that this isn't how things are done in Rails 😄 

I've ended up here because I wanted to make things more "railsy". I think the railsy way is to leverage "resources", in routing, controllers and models, even for things which aren't saved in the DB like this case.

I like the idea of only using the standard rails action, not adding custom resource actions if its possible to avoid it and using extra routes for custom URLS. This makes reading the routes file useful for people to work out whats available and what it's likely to do.

This has got messy because we're not storing the user_upgrade_request so the confirmation page (under #show) is always available. When the user hits /new, clicks continue without checking the met_requirements box, they are   redirected to the show route by the #create handler. Refreshing that, gives the confirmation screen.

I think you are right, it's probably better to get rid of show and create a custom resource action for the confirmation page. It's clearer than what I've ended up with.

Reply:
Yeah, I've tied myself in knots around this one.

When we started this project I was much more used to writing small javascript apps in express, where i really like explicit routing and handlers - it makes it clear exactly whats mounted on URLs. You can see that reflected in our routes file.

It's been pointed out to me that this isn't how things are done in Rails 😄 

I've ended up here because I wanted to make things more "railsy". I think the railsy way is to leverage "resources", in routing, controllers and models, even for things which aren't saved in the DB like this case.

I like the idea of only using the standard rails action, not adding custom resource actions if its possible to avoid it and using extra routes for custom URLS. This makes reading the routes file useful for people to work out whats available and what it's likely to do.

This has got messy because we're not storing the user_upgrade_request so the confirmation page (under #show) is always available. When the user hits /new, clicks continue without checking the met_requirements box, they are   redirected to the show route by the #create handler. Refreshing that, gives the confirmation screen.

I think you are right, it's probably better to get rid of show and create a custom resource action for the confirmation page. It's clearer than what I've ended up with.